File: ViewBase/Source/Others/RelayCommand.vb
Imports System.Windows.Input

Public Class RelayCommand
    Implements ICommand

    Private ReadOnly _Execute As Action(Of Object)
    Private ReadOnly _CanExecute As Func(Of Object, Boolean) = Function(parameters) True
    Private ReadOnly _OnError As Action(Of Exception)
    Private ReadOnly _Finally As Action

    Sub New(execute As Action(Of Object), canExecute As Func(Of Object, Boolean))
        _Execute = execute
        _CanExecute = canExecute
    End Sub

    Sub New(execute As Action(Of Object), canExecute As Func(Of Object, Boolean), onError As Action(Of Exception), Optional [finally] As Action = Nothing)
        _Execute = execute
        _CanExecute = canExecute
        _OnError = onError
        _Finally = [finally]
    End Sub

    Sub New(execute As Action(Of Object))
        _Execute = execute
    End Sub

    Public Event CanExecuteChanged As EventHandler Implements ICommand.CanExecuteChanged

    Public Sub InternalExecute(parameter As Object) Implements ICommand.Execute
        If CanExecute(parameter) Then
            Try
                _Execute(parameter)
            Catch ex As Exception
                If _OnError Is Nothing Then
                    Throw ex
                Else
                    _OnError(ex)
                End If
            Finally
                _Finally?.Invoke
            End Try
        End If
    End Sub

    Public Function CanExecute(parameter As Object) As Boolean Implements ICommand.CanExecute
        Return _CanExecute(parameter)
    End Function

    Sub RaiseCanExecuteChanged()
        RaiseEvent CanExecuteChanged(Me, New EventArgs)
    End Sub
End Class

File: BaseView/src/Others/RelayCommand.cs
using System;
using System.Windows.Input;

namespace Wells.BaseView
{
    public class RelayCommand : ICustomCommand
    {

        readonly Action<object> _Execute;
        readonly Func<object, bool> _CanExecute;
        readonly Action<Exception> _OnError;
        readonly Action _Finally;

        public event EventHandler CanExecuteChanged
        {
            add => CommandManager.RequerySuggested += value;
            remove => CommandManager.RequerySuggested -= value;
        }


        public RelayCommand(Action<object> execute, Func<object, bool> canExecute)
        {
            _Execute = execute;
            _CanExecute = canExecute;
        }

        public RelayCommand(Action<object> execute, Func<object, bool> canExecute, Action<Exception> onError)
        {
            _Execute = execute;
            _CanExecute = canExecute;
            _OnError = onError;
        }

        public RelayCommand(Action<object> execute, Func<object, bool> canExecute, Action<Exception> onError, Action onFinish)
        {
            _Execute = execute;
            _CanExecute = canExecute;
            _OnError = onError;
            _Finally = onFinish;
        }

        public RelayCommand(Action<object> execute)
        {
            _Execute = execute;
            _CanExecute = new Func<object, bool>((param) => true);
        }

        public void Execute(object parameter)
        {
            if (CanExecute(parameter))
            {
                try
                {
                    _Execute(parameter);
                }
                catch (Exception ex)
                {
                    if (_OnError == null)
                    {
                        throw;
                    }
                    else
                    {
                        _OnError(ex);
                    }
                }
                finally
                {
                    _Finally?.Invoke();
                }
            }
        }

        public bool CanExecute(object parameter)
        {
            return _CanExecute != null ? _CanExecute(parameter) : true;
        }

        private void OnCanExecuteChanged(object sender, EventArgs eventArgs)
        {
            CanExecute(eventArgs);
        }

        public void OnCanExecuteChanged()
        {
            CanExecute(null);
        }
      
    }

    public interface ICustomCommand : ICommand
    {
        void OnCanExecuteChanged();
    }
}
